website/blog/2020-02-24-presto-lambda.md

```diff
@@ -0,0 +1,193 @@
+
+## Lambda Execution
+
+In this section we are going to discuss how lambda is executed during runtime. The original implementation uses `MethodHandle` object to represent lambda on stack. Consider the same example:
+
+```sql
+transform(
+    elements, 
+    BIND(factor, (captured_factor, v) -> v * captured_factor)
+)
+```
+
+Each invocation of `transform` works in the following way: 
+1. Push the Java object represents elements on the stack. When `elements` is `array(integer)`, the corresponding Java stack type is `IntArrayBlock`.
+2. Push the `MethodHandle` object represents captured lambda to the stack, i.e. binding `factor` to `(captured_factor, v) -> v * captured_factor`. To this end: 
+    1. Push the `MethodHandle` object represents `v -> v * captured_factor` on the stack.
+    2. Push `captured_factor` on the stack.
+    3. Invoke `MethodHandle#bindTo` to get a `BoundMethodHandle` represents captured lambda on the top of stack. 
+3. Invoke `transform`. 
+
+Unfortunately, this implementation causes Java to generate a separate customized LambdaForm class for every `MethodHandle#bindTo` call (i.e. per each row). Such excessive runtime class generation quickly fills in metaspace full and causes full GC, see [Presto#7935](https://github.com/prestodb/presto/issues/7935) for reproduction and details. JDK developers also confirms that it is expected to have each `BoundMethodHandle` to be customized independently, and `MethodHandle#bindTo` [is not a good fit for implementing lambda capturing](http://mail.openjdk.java.net/pipermail/mlvm-dev/2017-May/006755.html).
+
+To fix this, we redesign Presto lambda execution via [Presto #8031](https://github.com/prestodb/presto/pull/8031). The key observations are:
+- Lambda capture has to be performed per invocation, as different value will be captured for each row).
+- However, we should use the same class representing captured lambda for every `BIND` call, otherwise we will generate too many classes.
+
+We use the same way as Java handles lambda and capture [3, 4]:
+- Lambda is represented as an object whose type is an functional interface (a.k.a. Single Abstract Method class)
+- The `invokedynamic` instruction is used to perform lambda capture:
+    * During the first `invokedynamic` call, the class represents captured lambda (which is a functional interface) and the method that performs capture will be generated. This step is also called linkage and will only be done once.
+    * Every `invokedynamic` call performs capture and returns an instance of the desired functional interface. 
+
+With this design, `BIND` function will always be fused together with the lambda generation step to generate captured lambda in one single step -- we cannot first generate an object that represents uncaptured lambda, and performs a separate partial application step. Note this implementation also doesn't allow more general higher-order function that returns a function as result.
```

Comment:
With this design, `BIND` function will always be fused together with the lambda generation step to generate captured lambda in a single step -- we cannot first generate an object that represents the uncaptured lambda, and then perform a separate partial application step. Note this implementation also doesn't allow more general higher-order function that returns a function as result.